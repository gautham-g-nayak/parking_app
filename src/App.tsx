import React, { useReducer } from "react";
import appReducer from "./context/AppReducer";
import { BrowserRouter, Navigate, Route, Routes } from "react-router-dom";
import ParkingLots from "./pages/ParkingLots";
import Payment from "./pages/Payment";
import SlotInput from "./pages/SlotInput";

const initContextData = { slots: {} };
const AppContext: any = React.createContext([]);

function App() {
  const [appData, dispatchAppData] = useReducer<any>(
    appReducer,
    initContextData
  );

  return (
    <AppContext.Provider value={[appData, dispatchAppData]}>
      <BrowserRouter>
        <Routes>
          <Route path="/" element={<SlotInput />} />
          <Route path="/parking-lots" element={<ParkingLots />} />
          <Route path="/payment" element={<Payment />} />
          <Route path="*" element={<Navigate to="/" />} />
        </Routes>
      </BrowserRouter>
    </AppContext.Provider>
  );
}

export default App;
export { AppContext };
