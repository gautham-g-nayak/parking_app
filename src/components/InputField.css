.inputField {
  width: 40%;
  padding: 15px 20px;
  border-radius: 5px;
  font-size: 16px;
  margin: 4px 10px;
  border-color: grey;
  border-width: 1px;
}

.inputField:focus {
  outline: none;
  box-shadow: 0 0 4px 3px var(--foregroundColor);
}
