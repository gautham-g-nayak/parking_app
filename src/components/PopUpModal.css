.modalBackground {
  width: 100vw;
  height: 100vh;
  background-color: rgba(0, 0, 0, 0.5);
  position: fixed;
  display: flex;
  justify-content: center;
  align-items: center;
}

.modalContainer {
  border-radius: 12px;
  background-color: white;
  box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
  display: flex;
  flex-direction: column;
  padding: 25px;
}

.footer {
    padding-top: 10px;
}
