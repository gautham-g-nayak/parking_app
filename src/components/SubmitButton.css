.submitButton {
  border: none;
  color: white;
  padding: 15px 40px;
  text-align: center;
  border-radius: 5px;
  text-decoration: none;
  font-size: 16px;
  margin: 4px 10px;
}
