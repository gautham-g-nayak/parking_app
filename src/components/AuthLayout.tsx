import { useContext } from "react";
import { Outlet, Navigate } from "react-router-dom";
import { AppContext } from "../App";

const AuthLayout = () => {
  const [appData] = useContext(AppContext);
  const length = Object.entries(appData.slots).length;
  return length === 0 ? (
    <Navigate to="/" replace />
  ) : (
    <Outlet />
  );
};

export default AuthLayout;