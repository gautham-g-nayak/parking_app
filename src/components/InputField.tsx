import "./InputField.css";
import { InputFieldProps } from "../types";

const InputField = ({
  onChange,
  placeHolder,
  type = "text",
}: InputFieldProps) => {
  return (
    <input
      className="inputField"
      placeholder={placeHolder}
      type={type}
      onChange={onChange}
      min="1"
    />
  );
};

export default InputField;
