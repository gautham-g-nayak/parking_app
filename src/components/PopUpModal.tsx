import TextField from "@mui/material/TextField";
import { useContext, useState } from "react";
import { actions } from "../actions";
import { AppContext } from "../App";
import "./PopUpModal.css";

function PopUpModal({ setOpenModalPopMenu, currentSpaceIndex }: any) {
  const [currentRegNo, setCurrentRegNo] = useState("");
  const { setSlot } = actions;
  const [appData, dispatchAppData] = useContext(AppContext);

  const handleClosePopMenu = () => setOpenModalPopMenu(false);

  const inputHandler = (reg: {
    target: { value: React.SetStateAction<string> };
  }) => {
    setCurrentRegNo(reg.target.value);
  };
  const allotSpace = () => {
    dispatchAppData(setSlot(currentSpaceIndex, currentRegNo, Date.now(), 0));
    handleClosePopMenu();
  };

  return (
    <div className="modalBackground">
      <div className="modalContainer">
        <div>Enter Register No:</div>
        <TextField
          id="outlined-basic"
          label="Register Number"
          variant="outlined"
          margin="normal"
          className="popUpInput"
          onChange={inputHandler}
        />
        <div className="footer">
          <button
            style={{ backgroundColor: "#0c7ca8", cursor: "pointer" }}
            onClick={handleClosePopMenu}
            className="submitButton"
          >
            CANCEL
          </button>
          <button
            className="submitButton"
            onClick={allotSpace}
            style={{ backgroundColor: "#0c7ca8", cursor: "pointer" }}
          >
            SUBMIT
          </button>
        </div>
      </div>
    </div>
  );
}

export default PopUpModal;
