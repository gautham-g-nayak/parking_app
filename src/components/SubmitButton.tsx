import { useNavigate } from "react-router-dom";
import "./SubmitButton.css";
import { SubmitButtonProps } from "../types";
import { actions } from "../actions";
import { AppContext } from "../App";
import { useContext } from "react";

const SubmitButton = ({
  label,
  data,
  to,
  buttonCondition = true,
}: SubmitButtonProps) => {
  const { setSlot } = actions;
  const [appData, dispatchAppData] = useContext(AppContext);
  const navigate = useNavigate();

  const submitHandler = () => {
    const lots = Array.from({ length: +data }, (v, i) => `P${i + 1}`);
    lots.forEach((lot) => {
      dispatchAppData(setSlot(lot, "", 0, 0));
    });
    navigate(to);
  };

  return (
    <button
      onClick={submitHandler}
      className="submitButton"
      style={{
        pointerEvents: buttonCondition ? "auto" : "none",
        backgroundColor: buttonCondition ? "var(--mainColor)" : "#babfc4",
      }}
    >
      {label}
    </button>
  );
};

export default SubmitButton;
