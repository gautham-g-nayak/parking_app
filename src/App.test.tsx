import { fireEvent, render, screen } from "@testing-library/react";
import App, { AppContext } from "./App";
import PopUpModal from "./components/PopUpModal";
import appReducer from "./context/AppReducer";
import InputField from "./components/InputField";
import SubmitButton from "./components/SubmitButton";
import { BrowserRouter } from "react-router-dom";
import AuthLayout from "./components/AuthLayout";
import Payment from "./pages/Payment";
import SlotInput from "./pages/SlotInput";
import ParkingLots from "./pages/ParkingLots";

test("renders learn react link", () => {
  render(<App />);
  const title = screen.getByText(/PARKING LOT MANAGEMENT/i);
  expect(title).toBeInTheDocument();
});

describe("testing the appReducer", () => {
  test("test set slot from appReducer", () => {
    const currentState = { slots: {} };
    const action = {
      type: "SET_SLOT",
      payload: {
        space: "P1",
        data: {
          registerNumber: "KA20M3040",
          timeIn: 1659451506175,
          timeOut: 1659459529644,
        },
      },
    };
    const newState = appReducer(currentState, action);
    const expectedState = {
      slots: {
        P1: {
          registerNumber: "KA20M3040",
          timeIn: 1659451506175,
          timeOut: 1659459529644,
        },
      },
    };
    expect(newState).toStrictEqual(expectedState);
  });
});

describe("test pop up modal component", () => {
  test("render pop up modal component", () => {
    render(<PopUpModal />);
    screen.debug();
  });

  test("cancel button", () => {
    const mockFunction = jest.fn();
    render(
      <PopUpModal currentSpaceIndex={"P1"} setOpenModalPopMenu={mockFunction} />
    );
    screen.debug();
    const cancelButton = screen.getByText("CANCEL");
    fireEvent.click(cancelButton);
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });

  test("submit button", () => {
    const mockFunction = jest.fn();
    const wrapper = ({ children }: any) => (
      <AppContext.Provider value={[{}, mockFunction]}>
        {children}
      </AppContext.Provider>
    );
    render(
      <PopUpModal currentSpaceIndex={"P1"} setOpenModalPopMenu={jest.fn} />,
      { wrapper }
    );
    screen.debug();
    const submitButton = screen.getByText("SUBMIT");
    fireEvent.click(submitButton);
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });
});

describe("test Input field component", () => {
  test("render Input field component", () => {
    const inputProps = {
      placeHolder: "label",
      onChange: jest.fn(),
      type: "text",
    };
    render(<InputField {...inputProps} />);
    screen.debug();
  });

  test("snapshot testing", () => {
    const inputProps = {
      placeHolder: "label",
      onChange: jest.fn(),
      type: "text",
    };
    const { container } = render(<InputField {...inputProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test("check if on change is called", () => {
    const inputProps = {
      placeHolder: "label",
      onChange: jest.fn(),
      type: "text",
    };
    render(<InputField {...inputProps} />);
    const inputElement = screen.getByDisplayValue("");
    screen.debug();
    fireEvent.change(inputElement, { target: { value: "Redux" } });
    expect(inputProps.onChange).toHaveBeenCalledTimes(1);
  });
});

test("submit button component", () => {
  const mockFunction = jest.fn();
  const wrapper = ({ children }: any) => (
    <AppContext.Provider value={[{}, mockFunction]}>
      {children}
    </AppContext.Provider>
  );
  const submitProps = {
    label: "SUBMIT",
    data: "1",
    to: "/parking-lots",
    buttonCondition: true,
  };
  render(
    <BrowserRouter>
      <SubmitButton {...submitProps} />
    </BrowserRouter>,
    { wrapper }
  );
  screen.debug();
  const submitButton = screen.getByText("SUBMIT");
  fireEvent.click(submitButton);
  expect(mockFunction).toHaveBeenCalledTimes(1);
});

test("render Auth component", () => {
  const data = {
    slots: {
      P1: {
        registerNumber: "KA20M3040",
        timeIn: 1659451506175,
        timeOut: 1659459529644,
      },
    },
  };
  const mockFunction = jest.fn();
  const wrapper = ({ children }: any) => (
    <AppContext.Provider value={[data, mockFunction]}>
      {children}
    </AppContext.Provider>
  );
  render(<AuthLayout />, { wrapper });
  screen.debug();
});

test("render slot input page", () => {
  const data = {
    slots: {
      P1: {
        registerNumber: "KA20M3040",
        timeIn: 1659451506175,
        timeOut: 1659459529644,
      },
    },
  };
  const mockFunction = jest.fn();
  const wrapper = ({ children }: any) => (
    <AppContext.Provider value={[data, mockFunction]}>
      {children}
    </AppContext.Provider>
  );
  render(
    <BrowserRouter>
      <SlotInput />
    </BrowserRouter>,
    { wrapper }
  );
  screen.debug();
});

describe("parking lot page", () => {
  test("render parking lots page", () => {
    const data = {
      slots: {
        P1: {
          registerNumber: "KA20M3040",
          timeIn: 1659451506175,
          timeOut: 1659459529644,
        },
      },
    };
    const mockFunction = jest.fn();
    const wrapper = ({ children }: any) => (
      <AppContext.Provider value={[data, mockFunction]}>
        {children}
      </AppContext.Provider>
    );
    render(
      <BrowserRouter>
        <ParkingLots />
      </BrowserRouter>,
      { wrapper }
    );
    screen.debug();
  });

  test("random button", () => {
    const data = {
      slots: {
        P1: {
          registerNumber: "KA20M3040",
          timeIn: 1659451506175,
          timeOut: 1659459529644,
        },
      },
    };
    const mockFunction = jest.fn();
    const wrapper = ({ children }: any) => (
      <AppContext.Provider value={[data, mockFunction]}>
        {children}
      </AppContext.Provider>
    );
    render(
      <BrowserRouter>
        <ParkingLots />
      </BrowserRouter>,
      { wrapper }
    );
    screen.debug();
    const submitButton = screen.getByText("ALLOCATE RANDOM SPACE");
    fireEvent.click(submitButton);
  });
});

describe("payemnt page", () => {
  test("render payment page", () => {
    const data = {
      slots: {
        P1: {
          registerNumber: "KA20M3040",
          timeIn: 1659451506175,
          timeOut: 1659459529644,
        },
      },
    };
    const mockFunction = jest.fn();
    const wrapper = ({ children }: any) => (
      <AppContext.Provider value={[data, mockFunction]}>
        {children}
      </AppContext.Provider>
    );
    render(
      <BrowserRouter>
        <Payment />
      </BrowserRouter>,
      { wrapper }
    );
    screen.debug();
  });

  test("payment taken button", () => {
    const data = {
      slots: {
        P1: {
          registerNumber: "KA20M3040",
          timeIn: 1659451506175,
          timeOut: 1659459529644,
        },
      },
    };
    const mockFunction = jest.fn();
    const wrapper = ({ children }: any) => (
      <AppContext.Provider value={[data, mockFunction]}>
        {children}
      </AppContext.Provider>
    );
    render(
      <BrowserRouter>
        <Payment />
      </BrowserRouter>,
      { wrapper }
    );
    screen.debug();
    const submitButton = screen.getByText("PAYMENT TAKEN");
    fireEvent.click(submitButton);
  });

  test("cancel button", () => {
    const data = {
      slots: {
        P1: {
          registerNumber: "KA20M3040",
          timeIn: 1659451506175,
          timeOut: 1659459529644,
        },
      },
    };
    const mockFunction = jest.fn();
    const wrapper = ({ children }: any) => (
      <AppContext.Provider value={[data, mockFunction]}>
        {children}
      </AppContext.Provider>
    );
    render(
      <BrowserRouter>
        <Payment />
      </BrowserRouter>,
      { wrapper }
    );
    screen.debug();
    const submitButton = screen.getByText("CANCEL");
    fireEvent.click(submitButton);
  });
});
