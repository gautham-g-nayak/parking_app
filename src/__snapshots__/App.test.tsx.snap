// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Input field component snapshot testing 1`] = `
<input
  class="inputField"
  min="1"
  placeholder="label"
  type="text"
/>
`;
